### ISSUE TITLE
node.surface() does not work as intended in version 0.996.1

### ISSUE THREAD

**Issue by rolzy** [SamuraiT/mecab-python3#19]
My Environment:
OS: Tried on both Ubuntu 16.04 and Windows 10 
Python: 3.6.3
mecab-python3: 0.996.1

Hi,

I recently found that node.surface is not working as intended in version 0.996.1. 

```
#coding:utf-8

import MeCab

# MeCabのインスタンス作成
mecab = MeCab.Tagger('-Ochasen')
mecab.parseToNode('')       # MeCab上の不具合で一度空で解析すると解消される

# 一行毎に解析開始
text = 'これはテストです。Mecabバージョン0.996.1'
node = mecab.parseToNode(text)

# 一単語毎に解析
while node:
    word_surface = node.surface
    print(word_surface)
    node = node.next
```

This produces the following result: 
![image](https://user-images.githubusercontent.com/10916857/48598110-f9d60680-e9ac-11e8-8254-0e4e5d702053.png)

However, on previous versions:
![image](https://user-images.githubusercontent.com/10916857/48598176-3b66b180-e9ad-11e8-9734-ba938b32dd02.png)

Thanks!

**Comment by polm** [comment]
I ran into this too. Looks like 5c23ea33724e67a might be related.

**Comment by honnibal** [comment]
Happy to update the code in spaCy if there's a new way we should write this. Thanks for your work!

**Comment by polm** [comment]
To clarify the nature of the change, this used to be the way to print each token in an input string:

```
import MeCab
tokenizer = MeCab.Tagger()

node = tokenizer.parseToNode('日本語ですよ').next

while node.posid != 0:
    print(node.surface)
    node = node.next
```

Now this seems to almost work:

```
import MeCab
tokenizer = MeCab.Tagger()

node = tokenizer.parseToNode('日本語ですよ').next

while node.posid != 0:
    print(node.surface[:node.length]) # XXX this is what changed
    node = node.next
```

The second form works for some dummy inputs like `a b c d`, but for realistic strings like `日本語ですよ` the `node.length` value is wrong. 

I'll keep looking into this, including doing the versioning tests.

**Comment by tomoyukih-fw** [comment]
**TL;DR:** Reverting 5c23ea3 + small tweak should fix the problem.

I found that [MeCab document](https://taku910.github.io/mecab/bindings.html) says the behavior of `surface` is changed in script language bindings:

> ただし, surface のみ, 単語そのものが返るように変更しています.

That is, the behavior of `surface` in the original MeCab library is actually returning the rest of sentence (as reported in this issue), not a single word (the behavior we expected). 

--------
**NOTE**: We can see that behavior in the following C++ program:

sample.cpp :
```c++
#include <iostream>
#include <mecab.h>
int main(int argc, char **argv) {
    char input[1024] = "これは動作テストです";

    MeCab::Tagger *tagger = MeCab::createTagger("");
    const MeCab::Node* node = tagger->parseToNode(input);
    for (; node; node = node->next) {
        std::cout << node->surface   << std::endl;
    }
    delete tagger;
}
```

Compile & Run:
```
$ g++ `mecab-config --cflags` sample.cpp  `mecab-config --libs` && ./a.out
これは動作テストです
は動作テストです
動作テストです
テストです
です
```
------

Behavior change of `surface` should be done in SWIG, and [the SWIG interface file in the original MeCab repository has a code for that](https://github.com/taku910/mecab/blob/master/mecab/swig/MeCab.i#L151):
```
char* mecab_node_t_surface_get(mecab_node_t *n) {
  char *s = new char [n->length + 1];
  memcpy (s, n->surface, n->length);
  s[n->length] = '\0';
  return s;
}
```

This code above was included in this repository as well, but was removed in 5c23ea3 for some reason. Maybe because of segfault as polm pointed out in the comment above. Segfault can be fixed by adding the code below:
```
...
%immutable mecab_node_t::surface;

/* ADD THIS */
%extend mecab_node_t {
  char *surface;
}

%extend MeCab::Tagger {
...
```

In conclusion, reverting 5c23ea3 + adding the code above should fix this issue. I would send a PR if I have some time to do that unless someone wants to do that.

**Comment by polm** [comment]
@tomoyukih-fw Thanks! I had tried adding the `mecab_node_t_surface_get` function back and couldn't figure out why it wasn't being picked up - the `extend` fixed it. I went ahead and created a PR. 

@zackw I know you're busy, but if you or another maintainer could take a look at this it'd be much appreciated.

**Comment by tomoyukih-fw** [comment]
@zackw Sadly I don't have any answer for that. I only found [the MeCab document for language bindings](https://taku910.github.io/mecab/bindings.html) that mentions the change for `node->surface`, and we can only guess why the original author decided to do that. I guess it's simply easier to use, as users of the library are interest in each word in most use cases, while having pointers or positions is more efficient in the internal calculation.

Anyway, thank a lot for your quick action, @zackw and @polm !

**Comment by digglife** [comment]
This is really confusing. FORK IT?

**Comment by zackw** [comment]
PR #22 has been merged. Thanks for your patience and your assistance, @polm @tomoyukih-fw.

**Comment by zackw** [comment]
Everyone affected by this bug, please try the release candidate of packaged wheels available from https://test.pypi.org/project/mecab-python3/0.996.2rc2/ , this bug should be corrected. Thank you everyone for your patience. We plan to make a new official release in the next couple of weeks.

**Comment by polm** [comment]
Checked the release candidate - it works! Thanks!
